### PR TITLE
Fixes #1044 - Memoize parameter file_path filters and de-quadratify folder refresh

### LINKED ISSUE #1044: [BUG] Performance issues when using parametrization on a larger report repository
### Library Version

1.1.0

### Python Version

3.11

### Operating System

Linux

### Authentication Method

Azure CLI (az login)

### What is the problem?

Deployments get a lot more slow when having a large respositories, with a bunch of files. In my case I had 58 semantic models and reports, where the reports were using the enhanced metadata format. That means that for a single report we could have over 400 json files.  

Functionally the output stayed correct, but runtime was inflated by redundant local scanning work. This made deploy cycles noticeably heavier in medium-to-large Fabric workspaces and created a scalability bottleneck in day-to-day development and CI.

When deploying the example of 58 objects, It was taking more then 2 hours and wasn't even finished.

### Steps to reproduce

Add objects to your workspace repository with a large number of files. The deployment is almost exponentially growing with each report you add.

### Expected behavior

Deployment performance should scale more predictably with actual work performed. Parameter filters should be evaluated with file-local context where possible, and repository refresh should avoid rediscovering the same filesystem structure repeatedly. The deployment lead time should be almost linear with the number of objects you are deploying.

### Actual behavior

The implementation effectively re-walked significant parts of the repository in paths that run frequently, leading to redundant I/O and path computation. Functional output remained correct, but execution time was inflated due to avoidable repeated traversal.

### Is there an ideal solution?

_No response_

### Additional context, screenshots, logs, error output, etc

_No response_

### PR DESCRIPTION
**Pull request by ayeshurun**
## Description

Deployments that use parametrization on large repositories slowed down super-linearly (effectively quadratic), to the point where a 58-item repo with enhanced-metadata reports (400+ JSON files each) could run for 2h+ without finishing (#1044). This PR removes the redundant filesystem traversals responsible for that growth.

Two independent problems were confirmed in the current code:

- **Primary hot loop.** During publishing, `_replace_parameters` runs once per definition file, and each run called `extract_parameter_filters` -> `process_input_path` for every `find_replace` / `key_value_replace` entry. For wildcard `file_path` filters, that re-globbed the entire repository every time. Total work scaled as `files x params x full-repo-glob`, which is the "almost exponential growth per report" the issue describes. The resolved filter set only depends on `(repository_directory, file_path filter)`, which is invariant across all files in a single publish, so it is now memoized on an instance-scoped cache (`_parameter_filter_path_cache`, following the existing `_item_attribute_cache` pattern). The repo is now globbed once per unique filter instead of once per file per entry.

- **Folder refresh.** `_refresh_repository_folders` nested `folder.rglob("*")` inside `root_path.rglob("*")`, an independent O(tree^2) traversal. It now walks each `.platform` folder's ancestors a single time (O(tree)) and produces the same set of repository folders.

The memoization helper resolves the path outside the lock (so threads with different filters do not serialize on the expensive glob) and stores under the lock, and it falls back to direct resolution when no real cache is present, preserving existing behavior for callers that pass a mock workspace.

Added tests cover the memoization contract (resolution runs once for many calls, and once per unique filter) and folder-refresh edge cases (`.children` containers excluded, root-level `.platform` produces no out-of-root paths). Full suite: 1039 passed, 6 skipped; ruff format and check clean on the changed files.

## Linked Issue (REQUIRED)

Fixes: #1044

### CODE REVIEW

File: src/fabric_cicd/_parameter/_utils.py [hunk 1]
```diff
@@ -406,11 +406,50 @@ def extract_parameter_filters(workspace_obj: FabricWorkspace, param_dict: dict)
     """Extracts the item type, name, and path filters from the parameter dictionary, if present."""
     item_type = param_dict.get("item_type")
     item_name = param_dict.get("item_name")
-    file_path = process_input_path(workspace_obj.repository_directory, param_dict.get("file_path"))
+    file_path = _resolve_filter_path(workspace_obj, param_dict.get("file_path"))
```
[the remark this answers is not in the document]
Comment (ayeshurun) [reply src/fabric_cicd/_parameter/_utils.py]: Good catch. Corrected the annotation to `tuple[Optional[str], Optional[str], Optional[list[Path]]]`, matching that `item_type`/`item_name` come from `dict.get()` (so `None` when absent) and `file_path` resolves via `process_input_path` to `list[Path] | None`.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: tests/test_parameter_utils.py [hunk 2]
```diff
@@ -633,6 +633,55 @@ def test_extract_parameter_filters(self, mock_workspace):
...
+
+        filters = [
+            {"file_path": "**/a.json"},
+            {"file_path": "**/b.json"},
+            {"file_path": ["**/a.json", "**/b.json"]},
+            {},  # None filter
+        ]
+
+        with mock.patch("fabric_cicd._parameter._utils.process_input_path") as mock_process:
+            mock_process.return_value = []
```
[the remark this answers is not in the document]
Comment (ayeshurun) [reply tests/test_parameter_utils.py]: Agreed. The mock now uses a `side_effect` that mirrors production: it returns `None` for a missing/empty `file_path` filter and concrete `list[Path]` values otherwise, so the empty-filter case is no longer misrepresented as `[]`. The per-unique-filter resolution count assertion still holds.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: .changes/unreleased/optimization-20260708-143554.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+kind: optimization
+body: Improve parameter file_path filter and repository folder refresh performance
```
Comment (shirasassoon) [inline .changes/unreleased/optimization-20260708-143554.yaml]: modify - `file_path` for better readability

Comment (ayeshurun) [reply .changes/unreleased/optimization-20260708-143554.yaml]: Updated the changelog body to wrap `file_path` in backticks and reorder for readability: "Improve performance of the `file_path` parameter filter and repository folder refresh". Pushed in 51b60b1.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

**Review by shirasassoon — APPROVED** [review]
